drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,309 @@
+            // Wait for the
+            const RX_EMPTY_BIT: u32 = 1 << 6;
+            if status & RX_EMPTY_BIT == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy(status));
+            // Initial busy-loop for faster response
+            if i >= 32 {
+                sleep_for(1);
+            }
+        }
+    }
+
+    /// Clears the FPGA's internal FIFOs
+    fn clear_fifos(&self) {
+        self.write_reg(reg::SPICR, 0x8080);
+    }
+
+    /// Wait until the SPI flash is idle
+    fn wait_flash_busy(&self, t: Trace) {
+        // Wait for the busy flag to be unset
+        while (self.read_flash_status() & 1) != 0 {
+            ringbuf_entry!(t);
+            sleep_for(1);
+        }
+    }
+
+    /// Reads the STATUS1 register from flash
+    fn read_flash_status(&self) -> u8 {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 1);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_STATUS_1);
+        self.wait_fpga_busy();
+        self.read_reg(reg::RX_FIFO).to_le_bytes()[0]
+    }
+
+    /// Sets the write enable flag in flash
+    fn flash_write_enable(&self) {
+        self.write_reg(reg::DATA_BYTES, 0);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::WRITE_ENABLE);
+        self.wait_fpga_busy();
+    }
+
+    /// Erases the 64KiB flash sector containing the given address
+    fn flash_sector_erase(&mut self, addr: u32) {
+        self.flash_write_enable();
+        self.write_reg(reg::DATA_BYTES, 0);
+        self.write_reg(reg::ADDR, addr);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::BLOCK_ERASE_64KB_4B);
+
+        // Wait for the busy flag to be unset
+        self.wait_flash_busy(Trace::SectorEraseBusy);
+    }
+
+    /// Reads data from the given address into a `BufWriter`
+    fn flash_read(
+        &mut self,
+        offset: u32,
+        dest: &mut dyn idol_runtime::BufWriter<'_>,
+    ) -> Result<(), ()> {
+        loop {
+            let len = dest.remaining_size().min(PAGE_SIZE_BYTES);
+            if len == 0 {
+                break;
+            }
+            self.clear_fifos();
+            self.write_reg(reg::DATA_BYTES, len as u32);
+            self.write_reg(reg::ADDR, offset);
+            self.write_reg(reg::DUMMY_CYCLES, 8);
+            self.write_reg(reg::INSTR, instr::FAST_READ_QUAD_OUTPUT_4B);
+            for i in 0..len.div_ceil(4) {
+                self.wait_fpga_rx();
+                let v = self.read_reg(reg::RX_FIFO);
+                let v = v.to_le_bytes();
+                for (j, byte) in v.iter().enumerate() {
+                    let k = i * 4 + j;
+                    if k < len {
+                        dest.write(*byte)?;
+                    }
+                }
+            }
```

Comment:
wondered a bit if this structure, which is repeated a few times, is easily generalizable to both the `flash_read` and `flash_write` methods, but honestly...trying to factor it out is probably more trouble than it's worth...

Reply:
also, it could maybe be nice if there was a comment explaining that this kinda weird loop is because we read bytes out of the FIFO in 32-bit words, and depending on the length to read, we might overshoot the requisite length and have to bail early? i was able to figure it out fairly quickly but it might not be totally obvious to everyone...

Reply:
**opinion:** It's not too difficult code to read or understand, but a quick comment wouldn't be out of place. Words are still generally easier to understand. Also I personally love the comment-rich coding style I often see in Oxide's code, so more comments! :)